### PR TITLE
[Datasets] Fix tensor extension string formatting (repr).

### LINKED ISSUE #25423: [Datasets] TensorArray printing bug (for float arrays)
### What happened + What you expected to happen

A pandas dataframe (or series) with a TensorArray column throws exception in __repr__() if the TensorArray element type is a float number (but works ok with an integer dtype inside the TensorArray).

```
import numpy as np
import pandas as pd
from ray.data.extensions import TensorArray

tint = TensorArray(np.arange(4).reshape(2,2))
# same as above but float
tfloat = TensorArray(np.arange(4).astype('float').reshape(2,2))

ser = pd.Series(tint)
r = ser.__repr__()
print(r)
# 0   [0 1]
# 1   [2 3]
# dtype: TensorDtype

ser = pd.Series(tfloat)
r = ser.__repr__() # throws exception
```

### Versions / Dependencies

ray.__version__, pd.__version__, np.__version__,sys.version_info

('1.12.1',
 '1.4.2',
'1.22.4',
 sys.version_info(major=3, minor=8, micro=11, releaselevel='final', serial=0))

(though I've noticed this for several months, since last year)
I'm not clear what the contract with pandas.ExtensionArray is, it could be the issue is with pandas.

### Reproduction script
```
import numpy as np
import pandas as pd
from ray.data.extensions import TensorArray

tint = TensorArray(np.arange(4).reshape(2,2))
# same as above but float
tfloat = TensorArray(np.arange(4).astype('float').reshape(2,2))

ser = pd.Series(tint)
r = ser.__repr__()
print(r)
# 0   [0 1]
# 1   [2 3]
# dtype: TensorDtype

ser = pd.Series(tfloat)
r = ser.__repr__() # throws exception
```
### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by clarkzinzow**
Fixes tensor extension string formatting, e.g. when invoking the DataFrame repr.

## Related issue number

Closes #25423 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by JiahaoYao** [comment]
nice, LGTM! @clarkzinzow

**Review by JiahaoYao — APPROVED** [review]

File: python/ray/air/util/tensor_extensions/pandas.py [hunk 1]
```diff
@@ -30,23 +30,150 @@
...
+            tpl = "{v}" if leading_space is False else " {v}"
+            fmt_values.append(tpl.format(v=_format(v)))
+        return fmt_values
+
+    return format_strings_slim(fmt_array, self.leading_space)
+
+
+_FORMATTER_ENABLED_ENV_VAR = "TENSOR_COLUMN_EXTENSION_FORMATTER_ENABLED"
+
+if os.getenv(_FORMATTER_ENABLED_ENV_VAR, "1") == "1":
```
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply python/ray/air/util/tensor_extensions/pandas.py:164]: Default is enable!

File: python/ray/air/util/tensor_extensions/pandas.py [hunk 2]
```diff
@@ -30,23 +30,150 @@
...
 except ImportError:
     # ABCIndexClass changed to ABCIndex in Pandas 1.3
     from pandas.core.dtypes.generic import ABCIndexClass as ABCIndex
-from pandas.core.indexers import check_array_indexer, validate_indices
-import pyarrow as pa
 
-from ray.util.annotations import PublicAPI
+
+#############################################
+# Begin patching of ExtensionArrayFormatter #
```
Comment (jianoaix) [inline python/ray/air/util/tensor_extensions/pandas.py:56]: Are those mostly ported from pandas? Add link to source if so?

File: python/ray/data/tests/test_dataset.py [hunk 3]
```diff
@@ -624,6 +623,23 @@ def test_tensor_array_array_protocol(ray_start_regular_shared):
     )
 
 
+def test_tensor_array_dataframe_repr(ray_start_regular_shared):
```
Comment (jianoaix) [inline python/ray/data/tests/test_dataset.py:626]: This really is not specific to dataset and a test_pandas.py under ray/air/util/tensor_extensions will likely make more sense (and a util without dedicated testing seems scary).

**Review by jianoaix — COMMENTED** [review]
LG!

File: python/ray/data/tests/test_dataset.py (hunk 3 above)
Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset.py:626]: Agreed, but there are a bunch of other tests in this module targeting the tensor extension, and we should have a dedicated PR that moves all of the tests over.

Can we do that in a follow-up PR? I don't want to block a PR that needs to land by the branch cut on a test code move that is lower priority and could happen afterwards.

Comment (jianoaix) [reply python/ray/data/tests/test_dataset.py:626]: We can start with this single test and take retrospective work as followup.

Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset.py:626]: Having test coverage split across multiple modules would be much worse IMO. I'll just move all of the tests in this PR so we can unblock this.

**Comment by clarkzinzow** [comment]
@jianoaix Moved the tests, please take a look, thanks!

**Review by jianoaix — APPROVED** [review]
Thanks!

**Pull request merged**